lib/topologies/server.js

```diff
@@ -924,27 +921,16 @@ Server.prototype.logout = function(dbName, callback) {
 /**
  * Authenticate using a specified mechanism
  * @method
- * @param {string} mechanism The Auth mechanism we are invoking
- * @param {string} db The db we are invoking the mechanism against
- * @param {...object} param Parameters for the specific mechanism
+ * @param {MongoCredentials} credentials Authentication credentials for login
  * @param {authResultCallback} callback A callback function
  */
-Server.prototype.auth = function(mechanism, db) {
-  var self = this;
-
-  if (mechanism === 'default') {
-    mechanism = getDefaultAuthMechanism(self.ismaster);
+Server.prototype.auth = function(credentials, callback) {
+  if (credentials.mechanism.toLowerCase() === 'default') {
+    credentials.mechanism = getDefaultAuthMechanism(this.ismaster);
```

Comment:
should credentials be mutated? I'm starting to feel like more of our types should be frozen, and immutable for their lifetime

Reply:
Part of the benefit of mutation is that you can resolve it only once. Otherwise, someone can hold on to stale credentials, and they will need to re-resolve the negotiation each time the credentials are presented.